base/abstractarray.jl

```diff
@@ -2272,53 +2273,69 @@ function _typed_hvncat(::Type{T}, dims::Tuple{Vararg{Int, N}}, row_first::Bool,
         len += cat_length(a)
     end
     outlen = prod(outdims)
-    outlen == 0 && ArgumentError("too few elements in arguments, unable to infer dimensions") |> throw
-    len == outlen || ArgumentError("too many elements in arguments; expected $(outlen), got $(len)") |> throw
+    outlen == 0 && throw(ArgumentError("too few elements in arguments, unable to infer dimensions"))
+    len == outlen || throw(ArgumentError("too many elements in arguments; expected $(outlen), got $(len)"))
 
     # copy into final array
-    A = Array{T, nd}(undef, outdims...)
+    A = cat_similar(as[1], T, outdims)
     # @assert all(==(0), currentdims)
     outdims .= 0
     hvncat_fill!(A, currentdims, outdims, d1, d2, as)
     return A
 end
 
-function _typed_hvncat(::Type{T}, shape::Tuple{Vararg{Tuple, N}}, row_first::Bool, as...) where {T, N}
+_typed_hvncat(T::Type, shape::Tuple{Tuple}, row_first::Bool, xs...) =
+    (length(shape[1]) == 0 && throw(ArgumentError("each level of `shape` argument must have at least one value"))) ||
+        _typed_hvncat_1d(T, shape[1][1], Val(row_first), xs...)
+
+function _typed_hvncat(::Type{T}, shape::NTuple{N, Tuple}, row_first::Bool, as...) where {T, N}
+    all(>(0), tuple((shape...)...)) || throw(ArgumentError("`shape` argument must consist of positive integers"))
+
     d1 = row_first ? 2 : 1
     d2 = row_first ? 1 : 2
-    shape = collect(shape) # saves allocations later
-    shapelength = shape[end][1]
-    lengthas = length(as)
-    shapelength == lengthas || throw(ArgumentError("number of elements does not match shape; expected $(shapelength), got $lengthas)"))
 
     # discover dimensions
     nd = max(N, cat_ndims(as[1]))
     outdims = zeros(Int, nd)
     currentdims = zeros(Int, nd)
     blockcounts = zeros(Int, nd)
+    shapev = Vector{Tuple{Vararg{Int}}}(undef, nd)
+    for i ∈ eachindex(shape)
+        shapev[i] = shape[i]
+    end
+    for i ∈ N+1:nd
+        shapev[i] = shapev[N]
+    end
     shapepos = ones(Int, nd)
 
-    for i ∈ eachindex(as)
+    length(shapev[end]) == 1 || throw(ArgumentError("last level of shape must be one integer equal to the input length"))
+    shapelength = shapev[end][1]
+    lengthas = length(as)
+    shapelength == lengthas || throw(ArgumentError("number of elements does not match shape; expected $(shapelength), got $lengthas)"))
+    all(!isempty, shapev) || throw(ArgumentError("each level of `shape` argument must have at least one value"))
+    sum(tuple((shape...)...)) == N * shapelength || throw(ArgumentError("all levels of `shape` argument must sum to the same value"))
+
+    @inbounds for i ∈ eachindex(as)
```

Comment:
it seems very unlikely that a block of code this size is entirely inbounds

Reply:
Alright, let's analyze:

`blockcounts`, `currentdims`, `shapepos`, `outdims`, `shapev` are created to be length `nd` and are all accessed by the iterating index `d` which runs from `1:nd`.

`as` is accessed by `i`, which is generated by `eachindex(as)`. It is also accessed by `i - 1`, but only if `i != 1`

That leaves accessing the elements of each tuple in `shapev`, which is determined by the incrementing `shapepos[d]`. `shapepos` is initialized with `ones`, and I have a check to ensure all elements of `shapev` have at least one element. The one threat is if shapepos[d] is incremented too far. I'll take a closer look at that with different unexpected inputs and create tests for them.

Reply:
Do you have any performance numbers for it on/off?

Reply:
Huh, I guess it's not doing too much. Guess I should just remove it.
```
const e = fill(1, 2, 1, ntuple(x -> 2, 18)...);
const f = fill(1, 2, 2, ntuple(x -> 2, 18)...);
using BenchmarkTools
#without
julia> @btime hvncat(((20, 10), (30,), (30,)), true, ntuple(x -> e, 20)..., ntuple(x -> f, 10)...);
#  2.928 s (23 allocations: 160.00 MiB)
#  2.781 s (23 allocations: 160.00 MiB)
#  2.811 s (23 allocations: 160.00 MiB)

#with
@btime hvncat(((20, 10), (30,), (30,)), true, ntuple(x -> e, 20)..., ntuple(x -> f, 10)...);
#  2.840 s (23 allocations: 160.00 MiB)
#  2.940 s (23 allocations: 160.00 MiB)
#  2.842 s (23 allocations: 160.00 MiB)
```